core/controlplane/config/templates/cloud-config-controller

```diff
@@ -1126,26 +1126,59 @@ write_files:
                     REGION=$(metadata dynamic/instance-identity/document | jq -r .region)
                     [ -n "${REGION}" ]
 
-                    while sleep 30; do
-                      STATE=$(asg describe-auto-scaling-instances --region "${REGION}" --instance-ids "${INSTANCE_ID}" | jq -r '.AutoScalingInstances[].LifecycleState')
-                      if [ ! "${STATE}" = Terminating:Wait ]; then
-                        continue
+                    # Not customizable (for now)
+                    SPOT_POLL_INTERVAL=5
+                    SG_POLL_INTERVAL=30
+
+                    # Used to control when to execute each termination check
+                    secs=0
+
+                    # Used to identify which was the source that requested the instance termination ('spot' or 'asg', for now)
+                    termination_source=''
+
+                    # Instance termination detection loop
+                    while sleep ${SPOT_POLL_INTERVAL}; do
```

Comment:
what do you think about splitting it into multiple wait loop, depending on where you run?

```
wait_on_asg() {
  echo "Awaiting ASG lifecycle hook for drain signal"
  while sleep 30; do
    state=$(asg describe-auto-scaling-instances --region "${REGION}" --instance-ids "${INSTANCE_ID}" | jq -r '.AutoScalingInstances[].LifecycleState')
    if [ "${state}" = Terminating:Wait ]; then
      break
    fi
done
}

wait_on_spotfleet() {
  echo "Awaiting spotfleet notification for drain signal"
  while sleep "${SPOT_POLL_INTERVAL}"; do
    http_status=$(curl -o /dev/null -w '%{http_code}' -sL http://169.254.169.254/latest/meta-data/spot/termination-time)
    if [ "${http_status}" -eq 200 ]; then
      break;
    fi
done
}

if [ "$(curl -o /dev/null -w '%{http_code}' -sL http://169.254.169.254/latest/meta-data/spot/)" -eq 200 ]; then
  termination_source=spot
  wait_func=wait_on_spotfleet
else
 termination_source=asg
 wait_func=wait_on_asg
fi

while ! $wait_func; do
 echo "[bug] wait func returned unexpected error, not draining." >&2
done

while true; do
  echo Node is terminating, draining it...
  ...
done
```